packages/core/src/browser/saveable.ts

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { Widget } from '@phosphor/widgets';
+import { Message } from '@phosphor/messaging';
+import { Event, MaybePromise, Key } from '../common';
+import { AbstractDialog } from './dialogs';
+
+export interface Saveable {
+    readonly dirty: boolean;
+    readonly onDirtyChanged: Event<void>;
+    save(): MaybePromise<void>;
+}
+
+export interface SaveableSource {
+    readonly saveable: Saveable;
+}
+
+export namespace Saveable {
+    export function isSource(arg: any): arg is SaveableSource {
+        return !!arg && ('saveable' in arg);
+    }
+    export function is(arg: any): arg is Saveable {
+        return !!arg && ('dirty' in arg) && ('onDirtyChanged' in arg);
+    }
+    export function get(arg: any): Saveable | undefined {
+        if (is(arg)) {
+            return arg;
+        }
+        if (isSource(arg)) {
+            return arg.saveable;
+        }
```

Comment:
Can we have explicit `return undefined`? Same below in `getDirty`.

Reply:
I think the type system is capable to catch it, if you change a return type to `Saveable` it will be an error